pkg/kubelet/dockertools/kube_docker_client.go

```diff
@@ -0,0 +1,471 @@
+	data, err := json.Marshal(src)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(data, dst)
+}
+
+// convertFilters converts filters to the filter type in engine-api.
+func convertFilters(filters map[string][]string) dockerfilters.Args {
+	args := dockerfilters.NewArgs()
+	for name, fields := range filters {
+		for _, field := range fields {
+			args.Add(name, field)
+		}
+	}
+	return args
+}
+
+// convertEnv converts data to a go-dockerclient Env
+func convertEnv(src interface{}) (*docker.Env, error) {
+	m := make(map[string]interface{})
+	if err := convertType(&src, &m); err != nil {
+		return nil, err
+	}
+	env := &docker.Env{}
+	for k, v := range m {
+		env.SetAuto(k, v)
+	}
+	return env, nil
+}
+
+func (k *KubeDockerClient) ListContainers(options docker.ListContainersOptions) (apiContainers []docker.APIContainers, err error) {
+	const operation = "list_containers"
+	defer recordOperation(operation, time.Now())
+	defer func() { recordError(operation, err) }()
+
+	containers, err := k.client.ContainerList(getDefaultContext(), dockertypes.ContainerListOptions{
+		Size:   options.Size,
+		All:    options.All,
+		Limit:  options.Limit,
+		Since:  options.Since,
+		Before: options.Before,
+		Filter: convertFilters(options.Filters),
+	})
+	if err != nil {
+		return
+	}
+	apiContainers = []docker.APIContainers{}
+	if err = convertType(&containers, &apiContainers); err != nil {
+		apiContainers = nil
+	}
+	return
+}
+
+func (d *KubeDockerClient) InspectContainer(id string) (container *docker.Container, err error) {
+	const operation = "inspect_container"
+	defer recordOperation(operation, time.Now())
+	defer func() { recordError(operation, err) }()
+
+	containerJSON, err := d.client.ContainerInspect(getDefaultContext(), id)
+	if err != nil {
+		// TODO(random-liu): Use IsErrContainerNotFound instead of NoSuchContainer error
+		if dockerapi.IsErrContainerNotFound(err) {
+			err = &docker.NoSuchContainer{ID: id, Err: err}
+		}
+		return
+	}
+	container = &docker.Container{}
+	if err = convertType(&containerJSON, container); err != nil {
+		container = nil
+	}
+	return
+}
+
+func (d *KubeDockerClient) CreateContainer(opts docker.CreateContainerOptions) (container *docker.Container, err error) {
+	const operation = "create_container"
+	defer recordOperation(operation, time.Now())
+	defer func() { recordError(operation, err) }()
+
+	config := &dockercontainer.Config{}
+	if err = convertType(opts.Config, config); err != nil {
+		return
+	}
+	hostConfig := &dockercontainer.HostConfig{}
+	if err = convertType(opts.HostConfig, hostConfig); err != nil {
+		return
+	}
+	resp, err := d.client.ContainerCreate(getDefaultContext(), config, hostConfig, nil, opts.Name)
+	if err != nil {
+		return
+	}
+	container = &docker.Container{}
+	if err = convertType(&resp, container); err != nil {
+		container = nil
+	}
+	return
+}
+
+func (d *KubeDockerClient) StartContainer(id string, _ *docker.HostConfig) (err error) {
```

Comment:
@vishh Yeah, the `HostConfig` in `StartContainer` is deprecated #20454, but the `DockerInterface` still has it. I'll remove this in the following refactoring PRs. :)